### PR TITLE
Feat: Allow specifying of arbitrary table properties as part of the model definition

### LINKED ISSUE #1139: Feat: bigquery require_partition_filter, partition_expiration_date and others

### PR DESCRIPTION
**Pull request by izeigerman**
Closes #1139

### CODE REVIEW

File: sqlmesh/core/model/meta.py [hunk 1]
```diff
@@ -158,6 +162,46 @@ def _columns_validator(
...
+    ) -> t.Optional[t.Dict[str, TableProperty]]:
+        if v is None or not isinstance(v, exp.Expression):
+            return v
+
+        dialect = values.get("dialect")
+
+        if isinstance(v, (exp.Tuple, exp.Array)):
+            eq_expressions: t.List[exp.Expression] = v.expressions
+        elif isinstance(v, exp.Paren):
+            eq_expressions = [v.this]
```
Comment (tobymao) [inline sqlmesh/core/model/meta.py:177]: v.unnest()

File: sqlmesh/core/model/meta.py [hunk 2]
```diff
@@ -158,6 +162,46 @@ def _columns_validator(
...
+            key = eq_expr.this.name
+            value_expr = eq_expr.expression
+            if isinstance(value_expr, exp.Boolean) or value_expr.is_string:
+                value = value_expr.this
+            elif value_expr.is_int:
+                value = int(value_expr.this)
+            elif value_expr.is_number:
+                value = float(value_expr.this)
+            else:
+                value = value_expr.copy()
```
Comment (tobymao) [inline sqlmesh/core/model/meta.py:198]: you're expecting an expression here?

do you want to use the python evaluator here to convert sql into python?

Comment (izeigerman) [reply sqlmesh/core/model/meta.py:198]: No, not really. I think I do want to preserve an expression. If someone does something like `TIMESTAMP('<timestamp>')` I want to pass it as is, don't I?

Comment (tobymao) [reply sqlmesh/core/model/meta.py:198]: then why are you converting anything at all?

Comment (izeigerman) [reply sqlmesh/core/model/meta.py:198]: To be consistent with scenario when the dictionary is constructed and passed directly from Python or dbt models.

Comment (tobymao) [reply sqlmesh/core/model/meta.py:198]: i don't think it makes a difference because you call convert anyways

Comment (tobymao) [reply sqlmesh/core/model/meta.py:198]: or just call convert in the validator, so when python models pass it in, it is always an expression type

File: sqlmesh/core/engine_adapter/spark.py [hunk 3]
```diff
@@ -249,11 +249,13 @@ def _create_table_properties(
         partitioned_by: t.Optional[t.List[exp.Expression]] = None,
         partition_interval_unit: t.Optional[IntervalUnit] = None,
         clustered_by: t.Optional[t.List[str]] = None,
+        table_properties: t.Optional[t.Dict[str, t.Any]] = None,
```
Comment (tobymao) [inline sqlmesh/core/engine_adapter/spark.py:252]: change to exp.Expression

File: sqlmesh/core/engine_adapter/spark.py [hunk 4]
```diff
@@ -265,6 +267,10 @@ def _create_table_properties(
                     this=exp.Schema(expressions=partitioned_by),
                 )
             )
+
+        for key, value in (table_properties or {}).items():
+            properties.append(exp.Property(this=key, value=exp.convert(value)))
```
Comment (tobymao) [inline sqlmesh/core/engine_adapter/spark.py:272]: remove the convert here and do it in the validator?

File: sqlmesh/core/engine_adapter/spark.py (hunk 3 above)
Comment (izeigerman) [reply sqlmesh/core/engine_adapter/spark.py]: done

File: sqlmesh/core/model/meta.py (hunk 2 above)
Comment (izeigerman) [reply sqlmesh/core/model/meta.py]: updated

File: sqlmesh/core/engine_adapter/spark.py (hunk 4 above)
Comment (izeigerman) [reply sqlmesh/core/engine_adapter/spark.py]: udpated

**Review by tobymao — APPROVED** [review]

**Pull request merged**